File: streampark-flink/streampark-flink-connector/streampark-flink-connector-clickhouse/src/main/scala/org/apache/streampark/flink/connector/clickhouse/internal/ClickHouseWriterTask.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.clickhouse.internal

import org.apache.streampark.common.util.Logger
import org.apache.streampark.flink.connector.clickhouse.conf.ClickHouseHttpConfig
import org.apache.streampark.flink.connector.failover.{FailoverWriter, SinkRequest}

import io.netty.handler.codec.http.HttpHeaderNames
import org.asynchttpclient.{AsyncHttpClient, ListenableFuture, Request, Response}

import java.util.concurrent.{BlockingQueue, ExecutorService, TimeUnit}

import org.apache.streampark.common.util.Implicits._
import scala.util.Try

case class ClickHouseWriterTask(
    id: Int,
    clickHouseConf: ClickHouseHttpConfig,
    asyncHttpClient: AsyncHttpClient,
    queue: BlockingQueue[SinkRequest],
    callbackService: ExecutorService)
  extends Runnable
  with AutoCloseable
  with Logger {

  @volatile var isWorking = false

  val failoverWriter: FailoverWriter =
    new FailoverWriter(clickHouseConf.storageType, clickHouseConf.getFailoverConfig)

  override def run(): Unit =
    try {
      isWorking = true
      logInfo(s"Start writer task, id = $id")
      while (isWorking || queue.nonEmpty) {
        val req = queue.poll(300, TimeUnit.MILLISECONDS)
        if (req != null) {
          send(req)
        }
      }
    } catch {
      case e: Exception =>
        logError("Error while inserting data", e)
        throw new RuntimeException(e)
    } finally {
      logInfo(s"Task id = $id is finished")
    }

  def send(sinkRequest: SinkRequest): Unit = {
    // ClickHouse's http API does not accept EMPTY request body
    if (sinkRequest.sqlStatement == null || sinkRequest.sqlStatement.isEmpty) {
      logWarn(s"Skip empty sql statement")
      return
    }

    val requests = buildRequest(sinkRequest)
    requests.foreach(
      request => {
        logDebug(s"Ready to fire request: $request")
        val whenResponse = asyncHttpClient.executeRequest(request)
        val callback = respCallback(whenResponse, sinkRequest)
        whenResponse.addListener(callback, callbackService)
      })
  }

  private def buildRequest(sinkRequest: SinkRequest): List[Request] = {
    logDebug(s"There is [${sinkRequest.sqlStatement.size}] statement(s) in SinkRequest ")
    // ClickHouse's http API does not accept multiple statements, so requests should be built by splitting statements
    sinkRequest.sqlStatement.filter(_.nonEmpty).map(
      statement => {
        val host = clickHouseConf.getRandomHostUrl
        val builder = asyncHttpClient
          .preparePost(host)
          .setRequestTimeout(clickHouseConf.timeout)
          .setHeader(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=utf-8")
          .setBody(statement)
        if (clickHouseConf.credentials != null) {
          builder.setHeader(HttpHeaderNames.AUTHORIZATION, "Basic " + clickHouseConf.credentials)
        }
        builder.build
      })
  }

  def respCallback(whenResponse: ListenableFuture[Response], sinkRequest: SinkRequest): Runnable =
    new Runnable {
      override def run(): Unit = {
        Try(whenResponse.get()).getOrElse(null) match {
          case null =>
            logError(
              s"""Error ClickHouseSink executing callback, params = $clickHouseConf,can not get Response. """)
            handleFailedResponse(null, sinkRequest)
          case resp if resp.getStatusCode != 200 =>
            logError(
              s"Error ClickHouseSink executing callback, params = $clickHouseConf, StatusCode = ${resp.getStatusCode} ")
            handleFailedResponse(resp, sinkRequest)
          case _ =>
        }
      }
    }

  /**
   * if send data to ClickHouse Failed, retry maxRetries, if still failed,flush data to
   * failoverStorage
   *
   * @param response
   * @param sinkRequest
   */
  def handleFailedResponse(response: Response, sinkRequest: SinkRequest): Unit = {
    if (sinkRequest.attemptCounter > clickHouseConf.maxRetries) {
      logWarn(
        s"""Failed to send data to ClickHouse, cause: limit of attempts is exceeded. ClickHouse response = $response. Ready to flush data to ${clickHouseConf.storageType}""")
      failoverWriter.write(sinkRequest)
      logInfo(
        s"Failover Successful, StorageType = ${clickHouseConf.storageType}, size = ${sinkRequest.size}")
    } else {
      sinkRequest.incrementCounter()
      logWarn(
        s"Next attempt to send data to ClickHouse, table = ${sinkRequest.table}, buffer size = ${sinkRequest.size}, current attempt num = ${sinkRequest.attemptCounter}, max attempt num = ${clickHouseConf.maxRetries}, response = $response")
      queue.put(sinkRequest)
    }
  }

  override def close(): Unit = {
    isWorking = false
    failoverWriter.close()
  }
}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-clickhouse/src/main/java/org/apache/streampark/flink/connector/clickhouse/internal/ClickHouseWriterTask.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.clickhouse.internal;
import org.apache.streampark.flink.connector.clickhouse.conf.ClickHouseHttpConfig;
import org.apache.streampark.flink.connector.failover.FailoverWriter;
import org.apache.streampark.flink.connector.failover.SinkRequest;
import org.asynchttpclient.*;
import io.netty.handler.codec.http.HttpHeaderNames;
import org.slf4j.Logger; import org.slf4j.LoggerFactory;
import java.util.List; import java.util.concurrent.*;
public class ClickHouseWriterTask implements Runnable, AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(ClickHouseWriterTask.class);
    private final int id; private final ClickHouseHttpConfig clickHouseConf;
    private final AsyncHttpClient asyncHttpClient; private final BlockingQueue<SinkRequest> queue;
    private final ExecutorService callbackService; private volatile boolean isWorking = false;
    private final FailoverWriter failoverWriter;
    public ClickHouseWriterTask(int id, ClickHouseHttpConfig conf, AsyncHttpClient client, BlockingQueue<SinkRequest> queue, ExecutorService cb) {
        this.id = id; this.clickHouseConf = conf; this.asyncHttpClient = client; this.queue = queue; this.callbackService = cb;
        this.failoverWriter = new FailoverWriter(clickHouseConf.storageType, clickHouseConf.getFailoverConfig());
    }
    @Override public void run() {
        try {
            isWorking = true; LOG.info("Start writer task, id = {}", id);
            while (isWorking || !queue.isEmpty()) {
                SinkRequest req = queue.poll(300, TimeUnit.MILLISECONDS);
                if (req != null) send(req);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        } catch (Exception e) {
            LOG.error("Error while inserting data", e);
            throw new RuntimeException(e);
        }
        finally { LOG.info("Task id = {} is finished", id); }
    }
    void send(SinkRequest sinkRequest) {
        List<String> stmts = sinkRequest.getSqlStatement();
        if (stmts == null || stmts.isEmpty()) { LOG.warn("Skip empty sql statement"); return; }
        for (String statement : stmts) {
            if (statement == null || statement.isEmpty()) continue;
            String host = clickHouseConf.getRandomHostUrl();
            BoundRequestBuilder builder = asyncHttpClient.preparePost(host)
                .setRequestTimeout(clickHouseConf.timeout)
                .setHeader(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=utf-8").setBody(statement);
            if (clickHouseConf.credentials != null) builder.setHeader(HttpHeaderNames.AUTHORIZATION, "Basic " + clickHouseConf.credentials);
            ListenableFuture<Response> future = asyncHttpClient.executeRequest(builder.build());
            future.addListener(() -> {
                try {
                    Response resp = future.get();
                    if (resp == null || resp.getStatusCode() != 200) handleFailedResponse(resp, sinkRequest);
                } catch (Exception e) { handleFailedResponse(null, sinkRequest); }
            }, callbackService);
        }
    }
    void handleFailedResponse(Response response, SinkRequest sinkRequest) {
        if (sinkRequest.getAttemptCounter() > clickHouseConf.maxRetries) {
            LOG.warn("Failed to send data to ClickHouse, flushing to {}", clickHouseConf.storageType);
            failoverWriter.write(sinkRequest);
        } else {
            sinkRequest.incrementCounter();
            try { queue.put(sinkRequest); } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
    }
    @Override public void close() { isWorking = false; failoverWriter.close(); }
}
